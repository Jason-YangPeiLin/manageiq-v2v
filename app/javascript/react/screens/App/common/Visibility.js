import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'patternfly-react';

// This component has no role on the page, and will be hidden from screen readers (role="presentation").
// Renders the children wrapped in visibility: hidden if the condition is met.
// Either way, the children take up space, visibility will not change the flow of the layout.
// The onKeyUp handler and role are used to satisfy eslint-plugin-jsx-a11y.
const Visibility = ({ children, hidden }) => (
  <span style={{ visibility: hidden ? 'hidden' : '' }} onKeyUp={noop} role="presentation">
    {children}
  </span>
);

Visibility.propTypes = {
  children: PropTypes.node,
  hidden: PropTypes.bool
};

export default Visibility;
