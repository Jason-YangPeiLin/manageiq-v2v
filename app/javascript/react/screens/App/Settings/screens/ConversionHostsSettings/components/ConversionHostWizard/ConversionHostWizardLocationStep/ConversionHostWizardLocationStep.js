import React from 'react';
import PropTypes from 'prop-types';
import { reduxForm, Field } from 'redux-form';
import { required } from 'redux-form-validators';
import { Form, Spinner } from 'patternfly-react';

import { BootstrapSelect } from '../../../../../../common/forms/BootstrapSelect';
import { V2V_TARGET_PROVIDERS, RHV, OPENSTACK } from '../../../../../../../../../common/constants';
import { FETCH_TARGET_COMPUTE_URLS } from '../../../../../../../../../redux/common/targetResources/targetResourcesConstants';

import { stepIDs } from '../ConversionHostWizardConstants';

class ConversionHostWizardLocationStep extends React.Component {
  componentDidUpdate(prevProps) {
    const { fetchTargetClustersAction, fetchTargetComputeUrls } = this.props;
    const prevDerivedProps = this.getDerivedProps(prevProps);
    const newDerivedProps = this.getDerivedProps();
    if (prevDerivedProps.selectedProviderType !== newDerivedProps.selectedProviderType) {
      fetchTargetClustersAction(fetchTargetComputeUrls[newDerivedProps.selectedProviderType]);
    }
  }

  getDerivedProps = (props = this.props) => {
    const { locationStepForm } = props;
    return {
      selectedProviderType: locationStepForm.values && locationStepForm.values.providerType,
      selectedProviderId: locationStepForm.values && locationStepForm.values.provider
    };
  };

  render() {
    const { providers, isFetchingTargetClusters, targetClusters } = this.props;
    const { selectedProviderType, selectedProviderId } = this.getDerivedProps();

    const availableProviderOptions = V2V_TARGET_PROVIDERS.filter(option =>
      providers.some(provider => provider.type === option.type)
    );
    const selectedProviderOption = availableProviderOptions.find(option => option.id === selectedProviderType);
    const providersFilteredBySelectedType =
      selectedProviderOption && providers.filter(provider => provider.type === selectedProviderOption.type);

    const targetComputeFilteredBySelectedProvider = targetClusters.filter(
      computeResource => computeResource.ems_id === selectedProviderId
    );

    const selectFieldBaseProps = {
      component: BootstrapSelect,
      labelWidth: 2,
      controlWidth: 9,
      inline_label: true,
      option_key: 'id',
      option_value: 'name',
      required: true,
      validate: [required()]
    };

    return (
      <Form className="form-horizontal">
        <Field
          {...selectFieldBaseProps}
          name="providerType"
          label={__('Provider Type')}
          options={availableProviderOptions}
        />
        {selectedProviderType && (
          <Spinner loading={isFetchingTargetClusters}>
            <Field
              {...selectFieldBaseProps}
              name="provider"
              label={__('Provider')}
              options={providersFilteredBySelectedType}
            />
            {selectedProviderType === RHV && (
              <Field
                {...selectFieldBaseProps}
                name="cluster"
                label={__('Cluster')}
                options={targetComputeFilteredBySelectedProvider}
                disabled={!selectedProviderId}
              />
            )}
            {selectedProviderType === OPENSTACK && (
              <Field
                {...selectFieldBaseProps}
                name="project"
                label={__('Project')}
                options={targetComputeFilteredBySelectedProvider}
                disabled={!selectedProviderId}
              />
            )}
          </Spinner>
        )}
      </Form>
    );
  }
}

ConversionHostWizardLocationStep.propTypes = {
  locationStepForm: PropTypes.object,
  providers: PropTypes.arrayOf(PropTypes.object),
  fetchTargetClustersAction: PropTypes.func,
  fetchTargetComputeUrls: PropTypes.object,
  isFetchingTargetClusters: PropTypes.bool,
  targetClusters: PropTypes.arrayOf(PropTypes.object)
};

ConversionHostWizardLocationStep.defaultProps = {
  locationStepForm: {},
  fetchTargetComputeUrls: FETCH_TARGET_COMPUTE_URLS
};

export default reduxForm({
  form: stepIDs.locationStep,
  destroyOnUnmount: false
})(ConversionHostWizardLocationStep);
