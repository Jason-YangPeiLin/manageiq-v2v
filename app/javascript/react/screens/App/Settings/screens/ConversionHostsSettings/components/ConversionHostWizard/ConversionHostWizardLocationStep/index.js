import { connect } from 'react-redux';
import ConversionHostWizardLocationStep from './ConversionHostWizardLocationStep';

import { fetchTargetClustersAction } from '../../../../../../../../../redux/common/targetResources/targetResourcesActions';

import { stepIDs } from '../ConversionHostWizardConstants';

const mapStateToProps = ({
  form,
  providers: { providers },
  targetResources: { isFetchingTargetClusters, targetClusters }
}) => ({
  locationStepForm: form[stepIDs.locationStep],
  providers,
  isFetchingTargetClusters,
  targetClusters
});

const mergeProps = (stateProps, dispatchProps, ownProps) => Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  { fetchTargetClustersAction },
  mergeProps
)(ConversionHostWizardLocationStep);
