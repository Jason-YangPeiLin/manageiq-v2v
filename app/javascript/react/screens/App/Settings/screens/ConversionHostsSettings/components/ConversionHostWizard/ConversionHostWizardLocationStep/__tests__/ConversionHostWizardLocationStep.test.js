import React, { cloneElement } from 'react';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import { reducer as formReducer } from 'redux-form';

import { generateStore } from '../../../../../../../common/testReduxHelpers';
import ConversionHostWizardLocationStep from '../ConversionHostWizardLocationStep';
import { RHV, OPENSTACK } from '../../../../../../../../../../common/constants';

describe('conversion host wizard location step', () => {
  const store = generateStore({ form: formReducer }, {});

  const ProviderWrapper = (
    { children, ...props } // eslint-disable-line react/prop-types
  ) => <Provider store={store}>{cloneElement(children, props)}</Provider>;

  const mockRhvProvider = {
    id: '1',
    name: 'Mock RHV',
    type: 'ManageIQ::Providers::Redhat::InfraManager'
  };

  const mockOspProvider = {
    id: '2',
    name: 'Mock OSP',
    type: 'ManageIQ::Providers::Openstack::CloudManager'
  };

  const mockRhvClusters = [
    { id: '1', name: 'Mock RHV Cluster 1', ems_id: '1' },
    { id: '2', name: 'Mock RHV Cluster 2', ems_id: '1' },
    { id: '3', name: 'Some other RHV cluster', ems_id: '3' }
  ];

  const mockOspTenants = [
    { id: '1', name: 'Mock OSP Project 1', ems_id: '2' },
    { id: '2', name: 'Mock OSP Project 2', ems_id: '2' },
    { id: '3', name: 'Some other OSP project', ems_id: '3' }
  ];

  const getBaseProps = () => ({
    selectedProviderType: null,
    selectedProviderId: null,
    providers: [mockRhvProvider, mockOspProvider],
    fetchTargetClustersAction: jest.fn(),
    fetchTargetComputeUrls: {
      [RHV]: '/api/mock/rhv/compute',
      [OPENSTACK]: '/api/mock/openstack/compute'
    },
    isFetchingTargetClusters: false,
    targetClusters: [],
    resetFormAction: jest.fn(),
    store
  });

  const shallowDive = jsx => {
    const connectWrapper = shallow(jsx); // <Connect(Form(ConversionHostWizardLocationStep))>
    const formWrapper = connectWrapper.dive(); // <Form(ConversionHostWizardLocationStep)>
    const componentWrapper = formWrapper.dive(); // <ConversionHostWizardLocationStep>
    return componentWrapper.dive(); // shallow-rendered ConversionHostWizardLocationStep
  };

  it('renders the redux-form wrapper correctly', () => {
    const component = shallow(<ConversionHostWizardLocationStep {...getBaseProps()} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly in the initial state with both providers present', () => {
    const component = shallowDive(<ConversionHostWizardLocationStep {...getBaseProps()} />);
    expect(component).toMatchSnapshot();
  });

  it('renders correctly in the initial state with only RHV providers present', () => {
    const component = shallowDive(
      <ConversionHostWizardLocationStep {...getBaseProps()} providers={[mockRhvProvider]} />
    );
    expect(component).toMatchSnapshot();
  });

  it('calls fetchTargetClustersAction when the selectedProviderType changes', () => {
    const baseProps = getBaseProps();
    const component = mount(
      <ProviderWrapper>
        <ConversionHostWizardLocationStep {...baseProps} />
      </ProviderWrapper>
    );
    expect(baseProps.fetchTargetClustersAction).toHaveBeenCalledTimes(0);
    component.setProps({ selectedProviderType: RHV });
    expect(baseProps.fetchTargetClustersAction).toHaveBeenCalledTimes(1);
    expect(baseProps.fetchTargetClustersAction).toHaveBeenCalledWith(baseProps.fetchTargetComputeUrls[RHV]);
  });

  it('does not call fetchTargetClustersAction when the selectedProviderType does not change', () => {
    const baseProps = getBaseProps();
    const component = shallowDive(<ConversionHostWizardLocationStep {...baseProps} selectedProviderType={RHV} />);
    expect(baseProps.fetchTargetClustersAction).toHaveBeenCalledTimes(0);
    component.setProps({ unrelatedProp: 'foo' });
    expect(baseProps.fetchTargetClustersAction).toHaveBeenCalledTimes(0);
  });

  it('renders correctly when loading clusters after selecting a provider type', () => {
    const component = mount(
      <ProviderWrapper>
        <ConversionHostWizardLocationStep {...getBaseProps()} selectedProviderType={RHV} isFetchingTargetClusters />
      </ProviderWrapper>
    );
    expect(component.find('Spinner').props().loading).toBeTruthy();
  });

  it('renders correctly after loading clusters but before selecting a provider', () => {
    const component = mount(
      <ProviderWrapper>
        <ConversionHostWizardLocationStep
          {...getBaseProps()}
          selectedProviderType={RHV}
          targetClusters={mockRhvClusters}
        />
      </ProviderWrapper>
    );
    expect(component.find('Spinner').props().loading).toBeFalsy();
    expect(
      component
        .find('Field[name="cluster"]')
        .first()
        .props().options
    ).toEqual([]);
  });

  it('renders correctly after selecting a RHV provider', () => {
    const component = mount(
      <ProviderWrapper>
        <ConversionHostWizardLocationStep
          {...getBaseProps()}
          selectedProviderType={RHV}
          targetClusters={mockRhvClusters}
          selectedProviderId="1"
        />
      </ProviderWrapper>
    );
    expect(
      component
        .find('Field[name="cluster"]')
        .first()
        .props()
        .options.map(option => option.id)
    ).toEqual(['1', '2']);
  });

  it('renders correctly after selecting an OSP provider', () => {
    const component = mount(
      <ProviderWrapper>
        <ConversionHostWizardLocationStep
          {...getBaseProps()}
          selectedProviderType={OPENSTACK}
          targetClusters={mockOspTenants}
          selectedProviderId="2"
        />
      </ProviderWrapper>
    );
    expect(
      component
        .find('Field[name="cluster"]')
        .first()
        .props()
        .options.map(option => option.id)
    ).toEqual(['1', '2']);
  });

  it('resets the hosts step when the cluster field changes', () => {
    const baseProps = getBaseProps();
    const component = mount(
      <ProviderWrapper>
        <ConversionHostWizardLocationStep
          {...baseProps}
          selectedProviderType={RHV}
          targetClusters={mockRhvClusters}
          selectedProviderId="1"
        />
      </ProviderWrapper>
    );
    component
      .find('Field[name="cluster"]')
      .first()
      .props()
      .onChange();
    expect(baseProps.resetFormAction).toHaveBeenCalledWith('conversionHostWizardHostsStep');
  });
});
