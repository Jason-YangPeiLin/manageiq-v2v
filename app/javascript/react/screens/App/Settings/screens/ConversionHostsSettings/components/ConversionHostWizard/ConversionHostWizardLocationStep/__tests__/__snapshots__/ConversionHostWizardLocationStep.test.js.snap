// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversion host wizard location step renders correctly in the initial state with both providers present 1`] = `
<Connect(Form(ConversionHostWizardLocationStep))
  _reduxForm={null}
  destroyOnUnmount={false}
  enableReinitialize={false}
  fetchTargetClustersAction={[MockFunction]}
  fetchTargetComputeUrls={
    Object {
      "openstack": "/api/mock/openstack/compute",
      "rhevm": "/api/mock/rhv/compute",
    }
  }
  forceUnregisterOnUnmount={false}
  form="conversionHostWizardLocationStep"
  getFormState={[Function]}
  isFetchingTargetClusters={false}
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  providers={
    Array [
      Object {
        "id": "1",
        "name": "Mock RHV",
        "type": "ManageIQ::Providers::Redhat::InfraManager",
      },
      Object {
        "id": "2",
        "name": "Mock OSP",
        "type": "ManageIQ::Providers::Openstack::CloudManager",
      },
    ]
  }
  pure={true}
  resetFormAction={[MockFunction]}
  selectedProviderId={null}
  selectedProviderType={null}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  submitAsSideEffect={false}
  targetClusters={Array []}
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;

exports[`conversion host wizard location step renders correctly in the initial state with only RHV providers present 1`] = `
<Connect(Form(ConversionHostWizardLocationStep))
  _reduxForm={null}
  destroyOnUnmount={false}
  enableReinitialize={false}
  fetchTargetClustersAction={[MockFunction]}
  fetchTargetComputeUrls={
    Object {
      "openstack": "/api/mock/openstack/compute",
      "rhevm": "/api/mock/rhv/compute",
    }
  }
  forceUnregisterOnUnmount={false}
  form="conversionHostWizardLocationStep"
  getFormState={[Function]}
  isFetchingTargetClusters={false}
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  providers={
    Array [
      Object {
        "id": "1",
        "name": "Mock RHV",
        "type": "ManageIQ::Providers::Redhat::InfraManager",
      },
    ]
  }
  pure={true}
  resetFormAction={[MockFunction]}
  selectedProviderId={null}
  selectedProviderType={null}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  submitAsSideEffect={false}
  targetClusters={Array []}
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;

exports[`conversion host wizard location step renders the redux-form wrapper correctly 1`] = `
<Hoc
  destroyOnUnmount={false}
  enableReinitialize={false}
  fetchTargetClustersAction={[MockFunction]}
  fetchTargetComputeUrls={
    Object {
      "openstack": "/api/mock/openstack/compute",
      "rhevm": "/api/mock/rhv/compute",
    }
  }
  forceUnregisterOnUnmount={false}
  form="conversionHostWizardLocationStep"
  forwardedRef={null}
  getFormState={[Function]}
  isFetchingTargetClusters={false}
  keepDirtyOnReinitialize={false}
  persistentSubmitErrors={false}
  providers={
    Array [
      Object {
        "id": "1",
        "name": "Mock RHV",
        "type": "ManageIQ::Providers::Redhat::InfraManager",
      },
      Object {
        "id": "2",
        "name": "Mock OSP",
        "type": "ManageIQ::Providers::Openstack::CloudManager",
      },
    ]
  }
  pure={true}
  resetFormAction={[MockFunction]}
  selectedProviderId={null}
  selectedProviderType={null}
  shouldAsyncValidate={[Function]}
  shouldError={[Function]}
  shouldValidate={[Function]}
  shouldWarn={[Function]}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  submitAsSideEffect={false}
  targetClusters={Array []}
  touchOnBlur={true}
  touchOnChange={false}
  updateUnregisteredFields={false}
/>
`;
